add scripts to build docker images - configure .gitlab-ci

tglatt opened the pull request.

Commit:
ci: add install-dependencies stage

Commit:
ci: add Dockerfile and .dockerignore

Commit:
ci: add build-and-push-tremplin-frontend-image job

Commit:
ci: push frontend docker image

Commit:
ci: add scripts building api image

Commit:
ci: add scripts building app image

Commit:
ci: update app dockerfile and add next.config.js

Commit:
ci: add docker-compose file and script to launch app

Commit:
ci: add build base stage

Commit:
ci: build base image

Commit:
ci: build base-dependencies image

Commit:
ci: build tremplin images

Commit:
ci: try running pipeline

Commit:
cis: attempt to build base image

Commit:
cis: attempt to build base image

Commit:
cis: attempt to build base image

Commit:
cis: attempt to build base image

Commit:
cis: attempt to build base image

Commit:
cis: attempt to build base image

Commit:
cis: attempt to build base image

Commit:
cis: attempt to build base image

Commit:
cis: attempt to build base image

Commit:
cis: attempt to build base image

Commit:
cis: attempt to build base image

Commit:
cis: attempt to build base-dependencies image

Commit:
cis: attempt to build base-dependencies image

Commit:
cis: attempt to build base-dependencies image

Commit:
cis: attempt to build base-dependencies image

Commit:
cis: attempt to build base-dependencies image

Commit:
cis: attempt to build base-dependencies image

Commit:
cis: attempt to build base-dependencies image

Commit:
cis: attempt to build tremplin images

Commit:
cis: attempt to build tremplin images

Commit:
cis: attempt to build tremplin images

Commit:
ci: update docker compose

Commit:
ci: update docker compose

Commit:
ci: update docker compose

Commit:
ci: clean script and docker file

Commit:
ci: test to add latest tag if branch name equals something

Commit:
ci: test to add latest tag if branch name equals something

Commit:
ci: test to add latest tag if branch name equals something

Commit:
ci: test to add latest tag if branch name equals something

Commit:
ci: test to add latest tag if branch name equals something

Commit:
ci: test to add latest tag if branch name equals something

Commit:
ci: test to add latest tag if branch name equals something

Commit:
ci: test to add latest tag if branch name equals something

Commit:
ci: test to add latest tag if branch name equals something

Commit:
ci: add latest tag if branch name equals master

Commit:
cis: delete useless Dockerfile

sboulkour commented on .gitlab-ci.yml:
```diff
@@ -1,11 +1,40 @@
+  - local: '/scripts/build/.gitlab-ci/build.gitlab-ci.yml'
+
+
+# build-and-push-tremplin-frontend-image:
```
Est-ce que cette partie commentée peut être virée ? :)

tglatt replied:
yes, je la vire

sboulkour commented on .gitlab-ci.yml:
```diff
@@ -1,11 +1,40 @@
+#   script:
+#   - cd packages/app
+#   - docker login  $CI_REGISTRY -u gitlab-ci-token -p $CI_JOB_TOKEN
+#   - docker build -t "$CI_REGISTRY_IMAGE/frontend:latest" .
```
"Latest is not a version" !
Est-ce que tagger avec le nom de la branche ou le SHA du commit ferai le job ? La CI fournie plein de variables qui devraient faire le job. Sur des builds en // exécutés depuis différentes branches, est-ce que ça risque pas de péter d'avoir uniquement `latest` ? Typiquement dans un cas où sur une branche il y a une MàJ des libs / dépendance et pas dans l'autre ?

tglatt replied:
Cette partie a été supprimée.

Sinon dans les autres scripts, les images sont marquées comme `latest` uniquement sur la branche `master`, sinon elles sont taggées avec comme ça: `$CI_COMMIT_REF_SLUG-$CI_COMMIT_SHA`.

sboulkour commented on scripts/build/.gitlab-ci/build.gitlab-ci.yml:
```diff
@@ -0,0 +1,35 @@
+    - apk update && apk add bash
+    - docker login  $CI_REGISTRY -u gitlab-ci-token -p $CI_JOB_TOKEN
+  after_script:
+    - docker logout
```
Très élégant le `docker logout` ! Je ne pense jamais à le faire et je crois que personne ne prend la peine de le faire d'habitude. Je kiffe.

tglatt replied:
thx

sboulkour commented on .gitlab-ci.yml:
```diff
@@ -1,11 +1,40 @@
+
+# build-and-push-tremplin-frontend-image:
+#   stage: "Build & Push Docker Image"
+#   image: docker:latest
```
Je pinnerai la version de docker (par ex 18.09) pour stabiliser le comportement du build dans le temps et ne pas avoir de mauvaise surprise. J'ajoute que l'environnement de run dans K8s, Docker est fixé sur 18.06 pour des questions de support.

tglatt replied:
https://github.com/SocialGouv/tremplin/commit/f24de0943114f100dc51acb1f994b150f0faff77

sboulkour commented on scripts/build/base-dependencies/base-dependencies.build.sh:
```diff
@@ -0,0 +1,90 @@
+do
+    case $i in
+        --version=*)
+            DOCKER_IMAGE_TAG="${i#*=}"
```
Qu'est-ce que ça fait le `#*=` juste après le nom de la variable ? J'ai jamais vu ça et Google n'a pas été avare d'explications :thinking:

tglatt replied:
le secret est ici: [Shell Parameter Expansion](https://www.gnu.org/software/bash/manual/html_node/Shell-Parameter-Expansion.html)

`${parameter#word}` renvoie `parameter` en virant la partie qui match la regexp `word`. Dans ce cas `*=` match `--version=`, donc ca renvoie le résultat la valeur de la version

sboulkour commented on .gitlab-ci.yml:
```diff
@@ -1,11 +1,40 @@
 ---
+image: node:10-alpine
```
Mismatch de version avec les Dockerfiles : `node:10.15-jessie` .
Question à cent balle, est-ce que ça serait utile de mettre le tag de l'image node à utiliser en variable ici, au début du fichier, et de la réutiliser partout (CI comme Dockerfiles) ?

tglatt replied:
https://github.com/SocialGouv/tremplin/commit/1c36fb074635a147924a400b6c5cc721dd161caa

sboulkour commented on scripts/build/.gitlab-ci/build.gitlab-ci.yml:
```diff
@@ -0,0 +1,35 @@
+.build-image:
+  image: docker:latest
```
Idem premier commentaire sur `docker:latest`

tglatt replied:
https://github.com/SocialGouv/tremplin/commit/f24de0943114f100dc51acb1f994b150f0faff77

sboulkour commented on .gitlab-ci.yml:
```diff
@@ -1,11 +1,40 @@
-    DOCKER_HOST: tcp://localhost:2375
+info:
+  stage: info
+  image: docker:latest
```
Idem premier commentaire sur `docker:latest`

tglatt replied:
https://github.com/SocialGouv/tremplin/commit/f24de0943114f100dc51acb1f994b150f0faff77

Commit:
ci: define default docker image - docker:18.09

ci: define docker image variable, use it in all stages

ci: define default docker image - docker:18.09

Commit:
cis: clean .gitlab-ci.yml

Commit:
ci: define docker node image variable, use it in ci and Dockerfile

douglasduteil commented on .dockerignore:
```diff
@@ -0,0 +1,2 @@
+node_modules
+packages/*/node_modules
```
already ignored due to previous line no ?

tglatt replied:
done

douglasduteil reviewed (changes requested):
You might want to use `docker-compose` instead of custom sh build scripts

tglatt committed:
chore: clean .dockerignore

douglasduteil commented:
Can be closed ?

tglatt commented:
yes, you can
